DynamicTablesPkg/Library/Common/AmlLib/CodeGen/AmlResourceDataCodeGen.c

```diff
@@ -1475,6 +1475,93 @@ AmlCodeGenRdRegister (
   return LinkRdNode (RdNode, NameOpNode, NewRdNode);
 }
 
+/** Code generation for the "IO ()" ASL function.
+
+  The Resource Data effectively created is a IO Resource
+  Data. Cf ACPI 6.5:
+   - s19.6.65 IO (IO Resource Descriptor Macro)
+   - s6.4.2.5 I/O Port Descriptor
+
+  The created resource data node can be:
+   - appended to the list of resource data elements of the NameOpNode.
+     In such case NameOpNode must be defined by a the "Name ()" ASL statement
+     and initially contain a "ResourceTemplate ()".
+   - returned through the NewRdNode parameter.
+
+  @param [in]  IsDecoder16          Decoder parameter.
+                                    TRUE if 16-bit decoder.
+                                    FALSE if 10-bit decoder.
+  @param [in]  AddressMinimum       Minimum address.
+  @param [in]  AddressMaximum       Maximum address.
+  @param [in]  Alignment            Alignment.
+  @param [in]  RangeLength          Range length.
+  @param [in]  NameOpNode           NameOp object node defining a named object.
+                                    If provided, append the new resource data
+                                    node to the list of resource data elements
+                                    of this node.
+  @param [out] NewRdNode            If provided and success,
+                                    contain the created node.
+
+  @retval EFI_SUCCESS               The function completed successfully.
+  @retval EFI_INVALID_PARAMETER     Invalid parameter.
+**/
+EFI_STATUS
+EFIAPI
+AmlCodeGenRdIo (
+  IN  BOOLEAN IsDecoder16,
+  IN  UINT16 AddressMinimum,
+  IN  UINT16 AddressMaximum,
+  IN  UINT8 Alignment,
+  IN  UINT8 RangeLength,
+  IN  AML_OBJECT_NODE_HANDLE NameOpNode, OPTIONAL
+  OUT AML_DATA_NODE_HANDLE  *NewRdNode  OPTIONAL
+  )
+{
+  EFI_STATUS                   Status;
+  EFI_ACPI_IO_PORT_DESCRIPTOR  IoDesc;
+  AML_DATA_NODE                *IoNode;
+
+  if (AddressMinimum > AddressMaximum) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if (Alignment != 0) {
+    /// check the alignment
+    if ((AddressMinimum & (Alignment - 1)) != 0) {

```

Comment:
I think this kind of check only works for addresses that are a multiple of 2.
Also if AddrMin=0x10 and Alignment=0x101, then the check will pass. (This is just in case someone gives random values).

I think this would be safer to do what the iasl compiler does, i.e. use modulos:
https://github.com/acpica/acpica/blob/1d9c129746c737b68e642f9f29a894e2700fbf46/source/compiler/aslresource.c#L273

Also I don't think we need to check that `(AddressMinimum + RangeLength)` is aligned as the spec doesn't mention it.

Reply:
Hi Pierre Gondois,
I have made the necessary changes based on the comments.
In my opinion, the previous implementation and this implementation are identical;
however, the latter is more readable.
Ideally, the code should also check for `(Alignment > AddressMinimum)`, as this would handle the random values mentioned above.
Since the specification did not mention this, I did not include it.
Thanks
AbduL